subsys/bluetooth/audio/mcc.c

```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+	    cur_mcs_inst->parent_group_obj_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->current_group_obj_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->current_group_obj_id_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
+
+#endif /* CONFIG_BT_MCC_OTS */
+
+	if (cur_mcs_inst->playing_order_handle &&
+	    cur_mcs_inst->playing_order_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->playing_order_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->playing_order_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
+
+	if (cur_mcs_inst->media_state_handle &&
+	    cur_mcs_inst->media_state_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->media_state_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->media_state_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
+
+	if (cur_mcs_inst->cp_handle &&
+	    cur_mcs_inst->cp_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->cp_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->cp_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
+
+	if (cur_mcs_inst->opcodes_supported_handle &&
+	    cur_mcs_inst->opcodes_supported_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->opcodes_supported_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->opcodes_supported_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
+
+#ifdef CONFIG_BT_MCC_OTS
+	if (cur_mcs_inst->scp_handle &&
+	    cur_mcs_inst->scp_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->scp_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->scp_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
+
+	if (cur_mcs_inst->search_results_obj_id_handle &&
+	    cur_mcs_inst->search_results_obj_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->search_results_obj_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->search_results_obj_id_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
+#endif /* CONFIG_BT_MCC_OTS */
+
+#ifdef CONFIG_BT_MCC_OTS
+	/* Start discovery of included services to find OTS */
+	discover_included(conn);
+#else
+	/* If OTS is not configured, discovery ends here */
+	if (mcc_cb && mcc_cb->discover_mcs) {
+		mcc_cb->discover_mcs(conn, 0);
+	}
+#endif /* CONFIG_BT_MCC_OTS */
```

Comment:
I believe all these services are tail-chained this way - at least vcs_client does more or less the same.
The discovery call triggers primary service discovery, which will trigger characteristic discovery (and subscription), which will trigger included service discovery, which will trigger discovery of characteristics of the included service.
And the callback is called when the chain ends.
The only difference here is that the subscriptions have been separated out as one additional step in the chain.

Reply:
The difference is that this is not a callback, but rather a specific function that is called to do something. Semantically that is different to me, and there's no good argument not to rename this to better match what it does IMO.

Reply:
I see your point - that is certainly a difference.
To me, it is not that important.  The function was anyway called from (a) callback(s), and fitted in the chain.
But I have reworked the thing so that the function has a more descriptive name - `subscribe_next_mcs_char` and does only that.
The actual tail-chaining is moved to the subscription callback, to fit the existing pattern.  (The subscription callback has also been renamed to fit the other callbacks in the chain.)